webapp/src/main/java/com/box/l10n/mojito/smartling/SmartlingClient.java

```diff
@@ -0,0 +1,188 @@
+        return headers;
+    }
+
+    HttpHeaders getHttpHeadersForJsonWithAuthorization() {
+        HttpHeaders headers = new HttpHeaders();
+        headers.setContentType(MediaType.APPLICATION_JSON);
+        headers.set("Authorization", "Bearer " + authToken);
+        return headers;
+    }
+
+    <T> HttpEntity<T> getHttpEntityForPayload(T payload) {
+        return new HttpEntity<>(payload, getHttpHeadersForJson());
+    }
+
+    <T> HttpEntity<T> getHttpEntityForPayloadWithAuthorization(T payload) {
+        return new HttpEntity<>(payload, getHttpHeadersForJsonWithAuthorization());
+    }
+
+    public void authenticate() throws SmartlingClientException {
+        Map<String, Object> payload = getPayloadMapAuthentication();
+        AuthenticationResponse authenticationResponse;
+        try {
+            authenticationResponse = postForAuthentication(payload, AUTH_API_V2_AUTHENTICATE);
+        } catch (SmartlingClientException e) {
+            throw new SmartlingClientException(e.getMessage());
+        }
+        if (authenticationResponse.isAuthenticationErrorResponse()) {
+            throw new SmartlingClientException("Error authenticating with Smartling API, please check your credentials");
+        } else if (!authenticationResponse.isSuccessResponse()) {
+            throw new SmartlingClientException(authenticationResponse.getErrorMessage());
+        } else {
+            Instant now = Instant.now();
+            lastAuthentication = now;
+            lastRefresh = now;
+            nextAuthentication = now.plus(Duration.ofHours(authenticationDurationHours));
+            nextRefresh = now.plus(Duration.ofMinutes(refreshDurationMinutes));
+            authToken = authenticationResponse.getResponse().getData().getAccessToken();
+            refreshToken = authenticationResponse.getResponse().getData().getRefreshToken();
+        }
+    }
+
+    AuthenticationResponse postForAuthentication(Map<String, Object> payload, String authApiV2) throws SmartlingClientException {
+        HttpEntity<Map<String, Object>> httpEntity = getHttpEntityForPayload(payload);
+        String postForObject = restTemplate.postForObject(
+                authApiV2, httpEntity, String.class);
+        AuthenticationResponse authenticationResponse;
+        try {
+            authenticationResponse = objectMapper.readValue(postForObject, AuthenticationResponse.class);
+        } catch (IOException e) {
+            logger.error(e.getMessage());
+            throw new SmartlingClientException(e.getMessage());
+        }
+        return authenticationResponse;
+    }
+
+    public void refresh() throws SmartlingClientException {
+        Instant now = Instant.now();
+        Instant nextCalculatedRefresh = now.plus(Duration.ofMinutes(refreshDurationMinutes));
+        if (nextRefresh.isAfter(now) && nextAuthentication.isAfter(now)) {
+            return;
+        } else if (nextCalculatedRefresh.isAfter(nextAuthentication)) {
```

Comment:
why is do we need `nextCalculatedRefresh`? Shouldn't `nextAuthentication` be enough to know when to do authentication again?

Reply:
this is to pre-emptively authenticate if this refresh is close to the next time we need to authenticate

Reply:
but will also add a check for if nextAuthentication is after now